### PR TITLE
Handle holes in boundaries using inner relation role

### PR DESCRIPTION
**Pull request by nlehuby**

### CODE REVIEW

File: src/boundaries.rs [hunk 1]
```diff
@@ -102,7 +102,55 @@ pub fn build_boundary(
...
     objects: &BTreeMap<osmpbfreader::OsmId, osmpbfreader::OsmObj>,
 ) -> Option<MultiPolygon<f64>> {
-    let roles = ["outer", "enclave", ""];
+    use geo::algorithm::contains::Contains;
+
+    let outer_polys = build_boundary_parts(relation, objects, vec!["outer", "enclave", ""]);
+    let inner_polys = build_boundary_parts(relation, objects, vec!["inner"]);
+    let mut multipoly = MultiPolygon(vec![]);
+
+    if inner_polys.is_none() {
```
Comment (antoine-de) [inline src/boundaries.rs]: it's more rustic to write this
```rust
if let inner_polys = Some(inner_polys) {
...
} else {
}
```
or 
```rust
match inner_poly {
    None => return outer_polys,
    Some(inner_polys) => {
        ....
    }
}
```

File: Cargo.toml [hunk 2]
```diff
@@ -6,4 +6,4 @@ authors = ["dt.ro <dt.ro@canaltp.fr>"]
 [dependencies]
 log = "0.4"
 osmpbfreader = "0.11"
-geo = "0.4.7"
+geo = "0.7.4"
```
Comment (amatissart) [inline Cargo.toml]: We should check if this upgrade does not break existing dependencies for cosmogony or mimirsbrunn.

File: src/boundaries.rs [hunk 3]
```diff
@@ -102,7 +102,55 @@ pub fn build_boundary(
...
+        return outer_polys;
+    }
+    let mut inner_polys = inner_polys.unwrap();
+
+    if outer_polys.is_some() {
```
Comment (antoine-de) [inline src/boundaries.rs]: I think you can rewrite this (not tested :stuck_out_tongue_winking_eye: ):

```rust
outer_polys.map(|outers|
    outers.map(|outer| {
        let mut i = 0;
        let mut inners = vec![];
        while(blablabla) {
            ...
        }
       Polygon::new(outer.exterior.clone(), inners)
    }).collect()
)

//this will already be an Option<Multipolygon>, you can return it directly
```

File: src/boundaries.rs [hunk 4]
```diff
@@ -102,7 +102,55 @@ pub fn build_boundary(
...
+        let outer_polys = outer_polys.unwrap();
+        for outer in outer_polys.0 {
+            let mut added_outer = false;
+            let mut i = 0;
+            let mut inners = vec![];
+            while i != inner_polys.0.len() {
+                let current_inner = inner_polys.0.get(i).unwrap().clone();
```
Comment (antoine-de) [inline src/boundaries.rs]: you shouldn't `clone` it there, you should clone it as late as possible

File: src/boundaries.rs [hunk 5]
```diff
@@ -102,7 +102,55 @@ pub fn build_boundary(
...
+                if outer.contains(&current_inner) {
+                    inner_polys.0.remove(i);
+                    inners.push(current_inner.exterior);
+                    added_outer = true;
+                } else {
+                    i += 1;
+                };
+            }
+            if !added_outer {
```
Comment (amatissart) [inline src/boundaries.rs]: `if inners.is_empty()` could do the trick, no ?

File: src/boundaries.rs [hunk 6]
```diff
@@ -381,3 +431,164 @@ fn test_build_two_boundary_closed() {
...
+            named_node(3.0, 3.0, "5"),
+            named_node(2.0, 3.0, "6"),
+            named_node(2.0, 2.0, "another_start"),
+        ])
+        .outer(vec![
+            named_node(4.0, 4.0, "yet_another_start"),
+            named_node(8.0, 4.0, "14"),
+            named_node(8.0, 8.0, "15"),
+            named_node(4.0, 8.0, "16"),
+            named_node(0.0, 0.0, "yet_another_start"),
```
Comment (amatissart) [inline src/boundaries.rs]: Same here.

File: src/boundaries.rs [hunk 7]
```diff
@@ -381,3 +431,164 @@ fn test_build_two_boundary_closed() {
...
+            named_node(2.0, 2.0, "12"),
+            named_node(1.0, 2.0, "13"),
+            named_node(1.0, 1.0, "other_start"),
+        ])
+        .outer(vec![
+            named_node(4.0, 4.0, "yet_another_start"),
+            named_node(8.0, 4.0, "4"),
+            named_node(8.0, 8.0, "5"),
+            named_node(4.0, 8.0, "6"),
+            named_node(0.0, 0.0, "yet_another_start"),
```
Comment (amatissart) [inline src/boundaries.rs]: It should be `4.0, 4.0`, I suppose.

File: src/boundaries.rs [hunk 8]
```diff
@@ -102,7 +102,55 @@ pub fn build_boundary(
...
+                    inners.push(current_inner.exterior);
+                    added_outer = true;
+                } else {
+                    i += 1;
+                };
+            }
+            if !added_outer {
+                multipoly.0.push(outer.clone());
+            } else {
+                multipoly
```
Comment (antoine-de) [inline src/boundaries.rs]: I don't know if that would'nt be too costly to cpoy those big geometry around, it might be better to change add the inner into the outer

**Review by amatissart — CHANGES_REQUESTED** [review]
That seems great !  
A few minor copy-paste errors in the tests.

File: src/boundaries.rs [hunk 9]
```diff
@@ -102,7 +102,55 @@ pub fn build_boundary(
...
+    let mut multipoly = MultiPolygon(vec![]);
+
+    if inner_polys.is_none() {
+        return outer_polys;
+    }
+    let mut inner_polys = inner_polys.unwrap();
+
+    if outer_polys.is_some() {
+        let outer_polys = outer_polys.unwrap();
+        for outer in outer_polys.0 {
```
Comment (amatissart) [inline src/boundaries.rs]: This algorithm does not handle the "donut-inside-donut" situation.  
A solution might be to sort `outer_polys` and `inner_polys` by area first ?

File: src/boundaries.rs [hunk 10]
```diff
@@ -102,7 +102,55 @@ pub fn build_boundary(
     relation: &osmpbfreader::Relation,
     objects: &BTreeMap<osmpbfreader::OsmId, osmpbfreader::OsmObj>,
 ) -> Option<MultiPolygon<f64>> {
-    let roles = ["outer", "enclave", ""];
+    use geo::algorithm::contains::Contains;
+
+    let outer_polys = build_boundary_parts(relation, objects, vec!["outer", "enclave", ""]);
+    let inner_polys = build_boundary_parts(relation, objects, vec!["inner"]);
+    let mut multipoly = MultiPolygon(vec![]);
```
Comment (antoine-de) [inline src/boundaries.rs]: it seems that this works (but for a test bug :confused: )

```rust
    if let Some(ref mut outers) = outer_polys {
        inner_polys.map(|inners| {
            inners.into_iter().for_each(|inner| {
                for ref mut outer in outers.0.iter_mut() {
                    if outer.contains(&inner) {
                        outer.interiors.push(inner.exterior);
                        break;
                    }
                }
            })
        });
    }
    outer_polys
```

Comment (nlehuby) [reply src/boundaries.rs]: I've changed the test to avoid the edge case with contains.

File: src/boundaries.rs (hunk 7 above)
Comment (nlehuby) [reply src/boundaries.rs]: I can change it if you want, but it works like this, because it is a named_node that is already defined.

**Review by antoine-de — APPROVED** [review]

File: Cargo.toml [hunk 11]
```diff
@@ -6,4 +6,4 @@ authors = ["dt.ro <dt.ro@canaltp.fr>"]
 [dependencies]
 log = "0.4"
 osmpbfreader = "0.11"
-geo = "0.7.4"
```
Comment (antoine-de) [inline Cargo.toml]: isn't it better to put `>0.6` ?

Comment (nlehuby) [reply Cargo.toml]: yes, right

Comment (nlehuby) [reply Cargo.toml]: hum, maybe not ... seems ok here, but Cosmogony and Mimir use 0.7.4 version with that syntax.

**Comment by nlehuby** [comment]
so many holes in Senlis

**Comment by nlehuby** [comment]
Successfully tested with Cosmogony https://github.com/osm-without-borders/cosmogony/pull/31 :tada:

**Review by amatissart — APPROVED** [review]
Let's ignore the "donut-inside-donut" and the "touching interiors" problems for the moment.
